### PR TITLE
[epg] Swap everything from 'std::map<unsigned int, CEpg*>' to 'EPGMAP'

### PR DESCRIPTION
**Pull request by AlwinEsch**
It was already defined in EpgContainer.h and now used on all places who have this type.

### CODE REVIEW

**Comment by AlwinEsch** [comment]
jenkins build this please

**Comment by Montellese** [comment]
Most of those cases could be handled with `auto` (especially the iterators).

**Comment by AlwinEsch** [comment]
Hi  @Montellese,
you mean for something like this
```C++
  for (auto &epg : m_epgs)
  {
    retval = epg.second->GetTag(iBroadcastId);
    if (retval)
      break;
  }
```

I can change it, a lot from C++11 is still new for me ;-)

**Comment by AlwinEsch** [comment]
How is CITR handled by auto?

**Comment by xhaggi** [comment]
`const auto&`

**Comment by AlwinEsch** [comment]
Thanks @xhaggi, easy with const on front ;-), know things like const_iterator and was not sure with them.

Is updated and use on all places from "EPGMAP" the ```auto```. Has checked it here and was OK, but better that it becomes checked also from your side :-)

**Comment by xhaggi** [comment]
looks good. @opdenkamp any objections?

File: xbmc/epg/EpgContainer.cpp [hunk 1]
```diff
@@ -245,16 +245,17 @@ bool CEpgContainer::PersistAll(void)
...
+  for (const auto &epgEntry : copy)
   {
-    CEpg *epg = it->second;
+    CEpg *epg = epgEntry.second;
     if (epg && epg->NeedsSave())
-    {
       bReturn &= epg->Persist();
-    }
+
+    if (m_bStop)
```
Comment (opdenkamp) [inline xbmc/epg/EpgContainer.cpp]: Fancy new notation has no advantages here with the condition moved out of the for statement. I'd keep the old notation here, so the break condition is defined in the for statement. and while you're at it, move the `CEpg* epg` declaration a few lines up so we don't create a new one in each iteration ;-)

File: xbmc/epg/EpgContainer.cpp [hunk 2]
```diff
@@ -340,25 +341,31 @@ CEpg *CEpgContainer::GetById(int iEpgId) const
...
 CEpgInfoTagPtr CEpgContainer::GetTagById(int iBroadcastId) const
 {
   CEpgInfoTagPtr retval;
   CSingleLock lock(m_critSection);
-  for (std::map<unsigned int, CEpg *>::const_iterator it = m_epgs.begin(); !retval && it != m_epgs.end(); ++it)
-    retval = it->second->GetTag(iBroadcastId);
+  for (const auto &epg : m_epgs)
+  {
+    retval = epg.second->GetTag(iBroadcastId);
+    if (retval)
```
Comment (opdenkamp) [inline xbmc/epg/EpgContainer.cpp]: same comment as above really, the new notation splits up the for statement from the break condition, which is meh. just change it to use `EPGMAP_CITR` if you want to improve readability

**Comment by opdenkamp** [comment]
mostly readability changes, no objections from me. but please keep the for statement together with the break condition instead of trying to force using the fancy new auto stuff. not only do you need more lines of code to write what you want to do, but it's also a lot cleaner.

File: xbmc/epg/EpgContainer.cpp (hunk 1 above)
Comment (Montellese) [inline xbmc/epg/EpgContainer.cpp]: This is not my code at all so feel free to ignore this comment but it took me a few minutes to figure out that the new code doesn't change behaviour (I actually already wrote a comment before I realized that it wasn't the case) because it is extremely easy to miss the `&& !m_bStop` in the for loop definition. The new code is much more obvious to read for me.

Comment (Montellese) [inline xbmc/epg/EpgContainer.cpp]: And either way you can still use `const auto&` instead of re-introducing `EPGMAP_CITR` even if you decide to go with the extended for loop syntax.

Comment (opdenkamp) [inline xbmc/epg/EpgContainer.cpp]: Well I got drilled into writing the break conditions into the for statement, so there's a single flow to follow without breaks in the middle that change it. So to me, readability is a lot better with everything in the for statement, than reading 3 times as many lines of code that does the same thing. This is an easy example, but in general, when loops get more complex, breaking in the middle and finding out what you're breaking out of and where you get into decreases readability imo.

And the resulting asm is a bit smaller too if there's a bunch of conditions checked in the single for statement than when you spread it out ;-)

Comment (xhaggi) [inline xbmc/epg/EpgContainer.cpp]: > because it is extremely easy to miss the && !m_bStop in the for loop definition. The new code is much more obvious to read for me.

I had the same discussion with @opdenkamp before and I prefer the break condition to be inside the loop, which improves readability a lot. But as said it's a matter of taste.

Comment (opdenkamp) [inline xbmc/epg/EpgContainer.cpp]: I admit that in this code, for 1 break statement, it doesn't change much for readability. But when you have more complex things like loops within loops, lots of code being executed within a loop, etc. then `break` in the middle becomes a problem, and code written like the original code becomes a lot cleaner to read. Especially when there's lots of these conditions.

And according to the prof that I had back then, it also increases the chance of somehow making it even more sloppy. Will spare you the details ;-)

Comment (opdenkamp) [inline xbmc/epg/EpgContainer.cpp]: So, in general, I avoid those nasty breaks in the middle

File: xbmc/epg/EpgContainer.cpp [hunk 3]
```diff
@@ -236,7 +236,7 @@ void CEpgContainer::LoadFromDB(void)
 bool CEpgContainer::PersistTables(void)
 {
   m_critSection.lock();
-  std::map<unsigned int, CEpg*> copy = m_epgs;
+  EPGMAP copy = m_epgs;
```
Comment (Paxxi) [inline xbmc/epg/EpgContainer.cpp]: Use auto here as well

File: xbmc/epg/EpgContainer.cpp (hunk 1 above)
Comment (AlwinEsch) [inline xbmc/epg/EpgContainer.cpp]: Hello @opdenkamp, tried several ways but has not found one to include it inside the "for". Has brought always a compile error, that's why I separated it.

How becomes it included in C++11 version?

Comment (opdenkamp) [inline xbmc/epg/EpgContainer.cpp]: if you use the `:` operator, then you can't. you'll have to use the standard for loop, but you can still use `auto` like this: `for (const auto& it = copy.begin(); it != copy.end() && !m_bStop; ++it)`

File: xbmc/epg/EpgContainer.cpp (hunk 2 above)
Comment (opdenkamp) [inline xbmc/epg/EpgContainer.cpp]: edit: this should have said `auto` instead of `EPGMAP_CITR` btw, zzz

File: xbmc/epg/EpgContainer.cpp (hunk 1 above)
Comment (AlwinEsch) [inline xbmc/epg/EpgContainer.cpp]: Has tried, also here was with auto always a error, using the old style
```for (EPGMAP::const_iterator it = m_epgs.begin(); !retval && it != m_epgs.end(); ++it)```

Comment (opdenkamp) [inline xbmc/epg/EpgContainer.cpp]: what's the error when you use auto? that should have worked afaik, though I'm not a c++11 expert (writing c99 mostly atm)

Comment (AlwinEsch) [inline xbmc/epg/EpgContainer.cpp]: comes with "auto &it", "const auto &it" and "const auto it" and was:
```
EpgContainer.cpp:251:70: error: passing ‘const std::_Rb_tree_iterator<std::pair<const unsigned int, EPG::CEpg*> >’ as ‘this’ argument of ‘std::_Rb_tree_iterator<_Tp>::_Self& std::_Rb_tree_iterator<_Tp>::operator++() [with _Tp = std::pair<const unsigned int, EPG::CEpg*>; std::_Rb_tree_iterator<_Tp>::_Self = std::_Rb_tree_iterator<std::pair<const unsigned int, EPG::CEpg*> >]’ discards qualifiers [-fpermissive]
   for (const auto it = copy.begin(); it != copy.end() && !m_bStop; ++it)
```

Randomly I has added this and has worked:
```C++
for (auto it = copy.begin(); it != copy.end() && !m_bStop; ++it)
```

Comment (xhaggi) [inline xbmc/epg/EpgContainer.cpp]: you can't use `const auto` as it makes the iterator `it` const not the element itself. so `++` and `--` won't work.  You can only use `for (auto it = ..)` but you need to const cast the vector/map otherwise the elements are non-const.

Example: "non-const" elements
```
std::vector<int> test = {1, 2, 3, 4};
for (auto it = test.begin(); it != test.end(); ++it)
  (*it) = 1;
```
Example: "const" elements
```
const std::vector<int> test = {1, 2, 3, 4};
for (auto it = test.begin(); it != test.end(); ++it)
  (*it) = 1; // won't work as element is const
```

Comment (xhaggi) [inline xbmc/epg/EpgContainer.cpp]: BTW it makes things really complicated and therefor you should use the forearch syntax with the break inside the loop or revert the changes to the `EPGMAP::const_iteration` for syntax like it was before.

**Comment by AlwinEsch** [comment]
Is updated and use the old style to have e.g. m_bStop inside "for".
```C++
for (EPGMAP::const_iterator it = copy.begin(); it != copy.end() && !m_bStop; ++it)
```

**Comment by AlwinEsch** [comment]
jenkins build this please

**Comment by AlwinEsch** [comment]
@opdenkamp, @xhaggi, OK to merge?

**Comment by opdenkamp** [comment]
looking good, thanks.

**Comment by MartijnKaijser** [comment]
@opdenkamp @xhaggi up to you if safe for RC1

**Comment by opdenkamp** [comment]
it's not fixing anything, so I'd wait

**Comment by AlwinEsch** [comment]
Not possible to do now? 
Want to include so much as possible from #6877 to only have the fault there and find them. Thought that this here is not a big change and comes in now.

**Comment by opdenkamp** [comment]
That PR did _a lot_ more than just fixing a segfault when I last 
checked, and imo it's a bad idea to change all of that at this stage, 
instead of just fixing a bug.
